text/0000-collection-view-redesign.md

```diff
@@ -0,0 +1,802 @@
+
+Staged operations will not be computed, instead they will add a new Transducer
+onto the operation stack.
+
+Forced operations will cause the current transducer stack to be run against the
+underlying collection, leading to an immediate value.
+
+It is a design decision whether or not to specially denote forced operations in
+the API.  While the current view API does not do this, it is because they are
+directly inheriting from the collections where each operations is always
+forced/performed immediately.   We'd like to propose that any operation
+which forces evaluation of a view be terminated with a uniform suffix, e.g. "!"
+
+
+## Implementation
+
+There is a [prototype library available][1] for those who would like to
+see this new implementation.   The library has implemented enough of the
+framework to prove:
+
+1. The implicit extension is viable, and can capture collection types
+   accurately.
+2. Storing a Builder is sufficient for ensuring the appropriate collection
+   is returned at the end of a chain of operations, according to the
+   `CanBuildFrom` logic.
+3. We can acheive sufficient (or better) performance than existing collection
+   views.
+4. Memoization is possible, and does not provide enough performance gain to
+   always be enabled.
+
+We believe this library could quickly be cleaned up and turned into a new
+Scala view module, while the existing lirbary be deprecated.  We would like to:
+
+1. Scala 2.12.x
+   - Include a scala-view module in the next 2.12 milestone released
+   - Mark existing `view` methods deprecated on all collections in scala master.
+2. Scala 2.13.x
+   - Remove existing View API from the standard library
+   - Expose custom Transducer library if there is demand.
+
+## Unresolved questions
+
+This proposal leaves open for discussion/contribution several things:
+
+1. Should operations which force evaluation have a common suffix (like `_!``) or
+   should they mirror the existing collections methods.
+2. More investigation into making the transducer/view design friendly for
+   parallel collections.
+3. Additional performance analysis/consideration for Transducer implementations
+   not in the prototype.
+4. Should `zip` force either of collections? It it worth it?
+5. What should be the signature of `flatMap`? should the argument function return `TraversableOnce` or a `View`?
+    [here](https://github.com/axel-angel/scala-blitzview/blob/master/src/main/scala/scala/collection/view/BlitzViews.scala#L21) @DarkDimius has a solution that allows to support both:
+
+   ```scala
+   def flatMap[C, U](f: El => U)(implicit ctx: Scheduler, viewable: BlitzView.IsViewable[U, C]): BlitzView[C]
+   ```
+
+## References
+
+1. [viewducers][1]
+2. [API Documentation][2]
+3. [Academic/Research papers/supporting material][3]
+4. [Alternative Libraries/Implementations][4]
+5. [Discussion forum/post/gitter/IRC][5]
+
+TODO - do we need more references?
+
```

Comment:
See also: https://gist.github.com/odersky/6b7c0eb4731058803dfd
Woudd be good to compare the code bases. In particular, I wondered how you do take/drop, because I was unsure about my implementation.